### PR TITLE
Divergency fix for QuickSelect

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 1]
```diff
@@ -183,21 +183,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
...
+                if (split_index < k)
+                    partition_start = split_index;
+            }
+            else {
+                auto rem_k = k - partition_start;
+                if (rem_k <= 0)
+                    break;
+                bool inside = local_id < partition_size;
+                auto offset = partition_start + local_id;
+                Float val = inside ? values[offset] : 1.0e30;
```
[the remark this answers is not in the document]
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:207]: You should probably use `constexpr` for this

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp]: Done

**Review by makart19 — APPROVED** [review]

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 2]
```diff
@@ -34,6 +34,8 @@ namespace oneapi::dal::backend::primitives {
 // Performs k-selection using Quick Select algorithm which is based on row partitioning
 template <typename Float>
 class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
+    static constexpr Float max_float = oneapi::dal::detail::limits<Float>::max();
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:37]: Just one usage, the function can be used directly in place

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 3]
```diff
@@ -183,21 +185,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
...
+                    bool is_mine = min_val == val && pos == -1;
+                    std::int32_t min_id = reduce(sg,
+                                                 is_mine ? local_id : local_size,
+                                                 sycl::ONEAPI::minimum<std::int32_t>());
+                    pos = min_id == local_id ? step : pos;
+                }
+                if (pos > -1) {
+                    values[partition_start + pos] = val;
+                    indices[partition_start + pos] = ind;
+                }
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:223]: Can you please extract a method?

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 4]
```diff
@@ -183,21 +185,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
...
+                                                                   pivot);
+                if ((split_index) == k)
+                    break;
+                if (split_index > k)
+                    partition_end = split_index;
+                if (split_index < k)
+                    partition_start = split_index;
+            }
+            else {
+                auto rem_k = k - partition_start;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:204]: remaining_k? It is good not to shorten variable names

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 5]
```diff
@@ -183,21 +185,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
...
+                if (rem_k <= 0)
+                    break;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:206]: Why this exit condition is valid for this branch only?

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 6]
```diff
@@ -183,21 +185,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
...
+                bool inside = local_id < partition_size;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:207]: Inside of what? Please give clear names to variables

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp [hunk 7]
```diff
@@ -183,21 +185,45 @@ class kselect_by_rows_quick : public kselect_by_rows_base<Float> {
...
+                auto offset = partition_start + local_id;
+                Float val = inside ? values[offset] : max_float;
+                std::int32_t ind = inside ? indices[offset] : -1;
+                std::int32_t pos = -1;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:211]: -1 is magic constant

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 2 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp:37]: Was discussed offline

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 3 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp]: Done

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 6 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp]: Renamed

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 4 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp]: Renamed

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 5 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp]: Removed

File: cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp (hunk 7 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_quick.hpp]: Fixed

**Review by michael-smirnov — APPROVED** [review]

**Pull request merged**